File: assertions/check_german.py
"""
Custom Python Assertion: Prüft ob die Antwort auf Deutsch ist
Verwendung in promptfooconfig.yaml:
  assert:
    - type: python
      value: file://assertions/check_german.py
"""

import re

def get_assert(output: str, context: dict) -> dict:
    """
    Prüft ob die Antwort auf Deutsch ist.
    
    Args:
        output: Die LLM-Antwort
        context: Kontext mit vars, prompt, etc.
    
    Returns:
        dict mit pass, score, reason
    """
    # Deutsche Umlaute und ß
    german_chars = re.findall(r'[äöüÄÖÜß]', output)
    
    # Typisch deutsche Wörter
    german_words = [
        'und', 'der', 'die', 'das', 'ist', 'sind', 'wird', 'werden',
        'bei', 'mit', 'für', 'auf', 'ein', 'eine', 'einer', 'eines',
        'nicht', 'auch', 'oder', 'aber', 'wenn', 'kann', 'können',
        'fahrzeug', 'lkw', 'truck', 'motor', 'achse'
    ]
    
    output_lower = output.lower()
    found_german_words = [w for w in german_words if w in output_lower]
    
    # Scoring
    has_umlauts = len(german_chars) > 0
    has_german_words = len(found_german_words) >= 3
    
    if has_umlauts and has_german_words:
        return {
            "pass": True,
            "score": 1.0,
            "reason": f"Antwort ist auf Deutsch (Umlaute: {len(german_chars)}, deutsche Wörter: {len(found_german_words)})"
        }
    elif has_german_words:
        return {
            "pass": True,
            "score": 0.8,
            "reason": f"Antwort enthält deutsche Wörter ({len(found_german_words)}), aber keine Umlaute"
        }
    else:
        return {
            "pass": False,
            "score": 0.0,
            "reason": "Antwort scheint nicht auf Deutsch zu sein"
        }

File: assertions/check_german.js
/**
 * JavaScript Assertion: Prüft ob Antwort auf Deutsch ist
 * 
 * Usage in Promptfoo:
 *   assert:
 *     - type: javascript
 *       value: file://assertions/check_german.js
 */

// Deutsche Wörter und Muster
const GERMAN_PATTERNS = [
  /\b(der|die|das|ein|eine|und|oder|ist|sind|wird|werden|hat|haben|kann|können)\b/gi,
  /[äöüß]/i,
  /\b(nicht|auch|aber|wenn|dann|weil|dass|damit|jedoch)\b/gi,
  /\b(Fahrzeug|Nummer|Zeichen|Hersteller|Information)\b/gi
];

// Englische Wörter (sollten nicht vorkommen)
const ENGLISH_PATTERNS = [
  /\b(the|is|are|was|were|has|have|can|will|would|should)\b/gi,
  /\b(vehicle|number|character|manufacturer|information)\b/gi
];

function getAssert(output, context) {
  if (!output || typeof output !== 'string') {
    return {
      pass: false,
      score: 0,
      reason: 'Keine Ausgabe vorhanden'
    };
  }

  let germanScore = 0;
  let englishScore = 0;

  // Zähle deutsche Muster
  for (const pattern of GERMAN_PATTERNS) {
    const matches = output.match(pattern);
    if (matches) {
      germanScore += matches.length;
    }
  }

  // Zähle englische Muster
  for (const pattern of ENGLISH_PATTERNS) {
    const matches = output.match(pattern);
    if (matches) {
      englishScore += matches.length;
    }
  }

  // Berechne Verhältnis
  const totalMatches = germanScore + englishScore;
  const germanRatio = totalMatches > 0 ? germanScore / totalMatches : 0;

  // Mindestens 70% deutsche Wörter
  const pass = germanRatio >= 0.7 && germanScore >= 3;
  const score = Math.min(germanRatio, 1);

  return {
    pass,
    score,
    reason: pass 
      ? `Antwort ist auf Deutsch (${Math.round(germanRatio * 100)}% deutsche Wörter)`
      : `Antwort ist nicht ausreichend auf Deutsch (${Math.round(germanRatio * 100)}% deutsche Wörter, min: 70%)`
  };
}

module.exports = getAssert;
